File: src/main/java/com/euc/grantfitassessment/eval/pipeline/ProgramAlignmentFilter.java
package com.euc.grantfitassessment.eval.pipeline;

import com.euc.core.EvaluationFilter;
import com.euc.core.EvaluationStage;
import com.euc.core.PipelineContext;
import com.euc.grantfitassessment.pipeline.GrantFitContextKeys;

/**
 * Scores EUC evaluation stage "programAlignment": the reasoned fit
 * classification must match ground truth. An eligibility failure
 * short-circuits fit reasoning entirely (docs/proposal.md Section 5), so
 * there is nothing to score here when the case is ineligible — it passes
 * vacuously rather than penalizing a case this criterion doesn't apply to.
 */
public class ProgramAlignmentFilter implements EvaluationFilter {

    @Override
    public Verdict evaluate(PipelineContext context, EvaluationStage stage) {
        boolean actualEligible = context.get(GrantFitContextKeys.ELIGIBLE, Boolean.class);
        if (!actualEligible) {
            return Verdict.PASSED;
        }

        String actualFit = context.get(GrantFitContextKeys.FIT_CLASSIFICATION, String.class);
        String expectedFit = context.get(GrantFitEvalContextKeys.EXPECTED_FIT_CLASSIFICATION, String.class);
        return expectedFit.equals(actualFit) ? Verdict.PASSED : Verdict.FAILED;
    }
}

File: src/euc/grantfitassessment/eval/pipeline/program_alignment_filter.py
"""EVAL-ALIGNMENT — checked against the assessment's own (actual) eligibility,
not the ground-truth eligibility: an ineligible verdict vacuously passes,
since fit reasoning never ran to produce one worth checking."""

from __future__ import annotations

from euc.core.context import PipelineContext
from euc.core.filters import Verdict
from euc.core.models import EvaluationCriterion
from euc.grantfitassessment import context_keys as prod_keys
from euc.grantfitassessment.eval import context_keys as eval_keys


def program_alignment_filter(context: PipelineContext, criterion: EvaluationCriterion) -> Verdict:
    actual_eligible: bool = context.get(prod_keys.ELIGIBLE)
    if not actual_eligible:
        return Verdict.PASSED

    actual_fit: str = context.get(prod_keys.FIT_CLASSIFICATION)
    expected_fit: str = context.get(eval_keys.EXPECTED_FIT_CLASSIFICATION)
    return Verdict.PASSED if expected_fit == actual_fit else Verdict.FAILED
